backend/open_webui/utils/user_encryption.py

```diff
@@ -0,0 +1,373 @@
+    """
+    
+    def __init__(self):
+        self.key_service_base_url = os.getenv(
+            'KEY_SERVICE_BASE_URL'
+        )
+        if not self.key_service_base_url:
+            raise Exception("KEY_SERVICE_BASE_URL is not set")
+
+        # In-memory cache for decrypted keys to avoid repeated API calls
+        self._key_cache: Dict[str, bytes] = {}
+    
+    def _generate_new_data_key(self) -> str:
+        """Generate a new encrypted data key from the external key service."""
+        url = f"{self.key_service_base_url}/keys"
+        response = requests.post(url, timeout=30)
+        response.raise_for_status()
+        
+        data = response.json()
+        if 'key' not in data:
+            raise Exception("Invalid response from key service")
+        
+        return data['key']
+    
+    def _decrypt_data_key(self, encrypted_key: str) -> str:
+        """Decrypt an encrypted data key using the external key service."""
+        url = f"{self.key_service_base_url}/keys/decrypt"
+        payload = {"key": encrypted_key}
+        response = requests.post(url, json=payload, timeout=30)
+        
+        if not response.ok:
+            error_text = response.text
+            raise Exception(f"Key service error {response.status_code}: {error_text}")
+        
+        data = response.json()
+        if 'decryptedKey' not in data:
+            raise Exception("Invalid response from key service")
+        
+        return data['decryptedKey']
+    
+    def _get_user_encryption_key(self, user_id: str) -> bytes:
+        """Get or create a user's encryption key as bytes for AES-256-GCM."""
+        # Check cache first
+        if user_id in self._key_cache:
+            return self._key_cache[user_id]
+        
+        # Check if user already has a data key in database
+        existing_key = UserDataKeys.get_user_data_key_by_user_id(user_id)
+        
+        if existing_key:
+            # User has an existing key, decrypt and cache it
+            decrypted_key_hex = self._decrypt_data_key(existing_key.encrypted_data_key)
+            key_bytes = bytes.fromhex(decrypted_key_hex)
+            
+            if len(key_bytes) != 32:
+                raise Exception("Invalid key length")
+            
+            self._key_cache[user_id] = key_bytes
+            return key_bytes
+        else:
+            # User doesn't have a key, generate a new one
+            encrypted_key = self._generate_new_data_key()
+            
+            # Store the encrypted key in the database
+            user_data_key = UserDataKeys.create_user_data_key(
+                user_id=user_id,
+                encrypted_data_key=encrypted_key
+            )
+            
+            if not user_data_key:
+                raise Exception("Failed to store user data key")
+            
+            # Decrypt the key before returning
+            decrypted_key_hex = self._decrypt_data_key(encrypted_key)
```

Comment:
LINE 115-122 is almost the same as LINE 93-100. maybe we can refactor the code a bit to reduce redundancy.